### DISCUSSION TITLE
$refs is undefined when used with Alpine store

### DISCUSSION THREAD

**Discussion by Arifursdev** [alpinejs/alpine#3604] [7. Bugs (archived)] [answered]
$refs is undefined when used with Alpine store as example below:
```
<html>
    <head>
        <script>
            document.addEventListener('alpine:init', function () {

                Alpine.store('test_store', {
                    text: 'Hello There',
                    init() {
                        const input = this.$refs.myInputElement;
                        console.log(input); // undefined
                    }
                })

            });
        </script>
        <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
    </head>

    <body>
        <div x-data x-text="$store.test_store.text"></div>
        <br>
        <input type="text" x-data x-ref="myInputElement">
    </body>
</html>
```

**Comment by ekwoka** [comment, marked as answer]
Yes. Because stores can't have refs.

Components have refs.

**Reply by Arifursdev** [reply]
Ah I see! 🤦‍♂️ strange, could be great if there was something similar for the store, instead of writing that querySelector(). 
Thanks for your quick response though. Appreciate it. 👍

**Reply by ekwoka** [reply]
You can assign the item to state in the store.

Or modelt he values.

Generally, refs are not the solution regardless.

Like here, it seems like you should just `x-model` the input to state in the store


No need to ever touch the element.

**Reply by Arifursdev** [reply]
I know. but this is just one example. there are other use cases with other elements, which if refs was available, it could be much more cleaner and lesser code and faster to do. I hate to ask and will never ask to make refs available in store.

imho, if team wishes I'm sure it would be much better as currently its bit awkward that devs can use store but you can't use refs. 🙂
 
I'm have not went further to know what else is not supported with store. perhaps it could be removed from the main file and have it as a separate plugin or something so devs can include it if they wish to use it. 🙂

**Reply by SimoTod** [reply]
If you think about that, the store doesn't have any html, it's just a shared container of data so $refs wouldn't be able to point to any element.
Because the store is meant to be shared, you can't bind it to a specific element. For the same reason, you won't have access to $el, $root, etc. 
Alpine preferes a declarative approach over an imperative one. You never set or manipulate html content/element. You update the state and let the html react.

**Reply by ekwoka** [reply]
You can also pretty simply do

```html
x-init="$store.myStore.targetEl = $el"
```

Or

```html
x-init="$store.myStore.setup($el)"
```

If it's really needed.

I've made a lot of Alpine apps, and plugins, from Ecommerce to fintech and virtually every time I've seen an x-ref used, there was a better way to do it that didn't use any refs.

Even when it's just directly calling a function with the element as an arg.

**Reply by Arifursdev** [reply]
that's a great idea! don't remember using `x-init` on element before.. but that again is not available on `init()` function of the store.
example:

I think with this x-init on element, the passed `$el` is not available on the store `init()` function:
```html
<html>
    <head>
        <script>
            document.addEventListener('alpine:init', function () {

                Alpine.store('test_store', {
                    text: 'Hello world',
                    myInputElement: null,
                    test($el){
                      this.myInputElement = $el // ✅ works
                      this.myInputElement.value = 'hey there'
                      this.someotherfun($el)
                    },
                    someotherfun(test$el){
                      console.log(test$el) // ✅ <input type="text" x-data="" x-init="$store.test_store.test($el)">
                    },
                    init() {
                      this.myInputElement.value = 'init' // ❌ this.myInputElement is null here
                    }
                })

            });
        </script>
        <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
    </head>

    <body>
        <div x-data x-text="$store.test_store.text"></div>
        <br>
        <input type="text" x-data x-init="$store.test_store.test($el)">
    </body>
</html>
```

also `$nextTick` is also not possible to use on this `Alpine.store()`.

**Reply by Arifursdev** [reply]
> 

for a good jQuery replacement, I think in the `Alpine.store()` I'm looking for helper functions for $refs and other things that will work with store.
like: 
```html
<div class="test" x-data x-store-ref="testElement"></div>
```
and in the store:
```js
init(){
    this.$storeRefs.testElement.innerText = 'test' // for store
}
```

in this case, so i don't have to write vanilla js querySelector and then check for its existence.

**Reply by SimoTod** [reply]
But the question is always the same why? 
You usually create your variable or getter in the store and use x-text instead of x-ref so it stays in sync no matter how you update the underlying data. 

If you need to write code where you manually set values in that way, then Jquery is ironically a better tool for the job.

The other question is, are you using the store for a single component? 
In that case, what you really want is Alpine.data which is the js representation of the component and has access to all the magics.

**Reply by Arifursdev** [reply]
> But the question is always the same why? You usually create your variable or getter in the store and use x-text instead of x-ref so it stays in sync no matter how you update the underlying data.
> 
> If you need to write code where you manually set values in that way, then Jquery is ironically a better tool for the job.
> 
> The other question is, are you using the store for a single component? In that case, what you really want is Alpine.data which is the js representation of the component and has access to all the magics.

one of the reason I like to use store because I can use it on anyplace on the DOM, and for readability (which store I'm using..) etc.
I know every case I have is doable without the `refs`. but I'm looking to use `refs` for the purpose of `refs`. if everything is doable without refs, why do we have refs, YKWIM.

I guess we all got the answer that its not possible to use refs in store in alpine js.

**Reply by ekwoka** [reply]
Refs exist almost entirely for integration with third party stuff, and i think may even only exist really for that out of some kind of convention.

Much like x-html shouldn't really be used but some people see it as required.


As I mentioned, if you have that function that adds the element to to store, just have the thing that modifies it be right there with it.

You don't need or really want the stores init to do that.

It's very round peg square hole.
